### ISSUE TITLE
Inference example on keras's IMDB dataset

### ISSUE THREAD

**Issue by taless474** [STEllAR-GROUP/keras#9]
To get the predicted labels on the test data, you can train a network using either of:
```python
model = Sequential()
model.add(Embedding(max_features, n_dim))
model.add(LSTM(128, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=15,
          validation_data=(x_test, y_test))
```
or
```python
model = Sequential()
model.add(Embedding(max_features, n_dim, input_length=maxlen))
model.add(SpatialDropout1D(dropout))
model.add(Conv1D(256, 3, activation='relu'))
model.add(GlobalMaxPool1D())
model.add(Dense(256, activation='relu'))
model.add(Dropout(dropout))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model.fit(x_train, y_train, 
          batch_size=batch_size, 
          epochs=epochs, 
          verbose=1, 
          validation_split=.50)
```
and run
```python
labels_pred = model.predict_proba(x_test)
```
Having the predicted labels, this is our inference example:
```python
from keras import backend as K
from keras.datasets import imdb
import numpy as np
import pandas as pd
from phylanx import Phylanx
import matplotlib.pyplot as plt

@Phylanx
def unique_eager(x):
    return np.unique(x)

unique = Phylanx.lazy(unique_eager)

@Phylanx
def argsort_eager(x):
    return np.argsort(x)

argsort = Phylanx.lazy(argsort_eager)

@Phylanx
def where_eager(x):
    return np.where(x)

where = Phylanx.lazy(where_eager)


max_features = 10000
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
largest_index = y_test.size - 1

df = pd.read_csv('labels_pred.csv')

labels_pred = K.squeeze(df.values, 1)

print("Predicted labels:", K.get_value(labels_pred))
plt.hist(K.get_value(labels_pred))
plt.show()

###
corrects = K.less(K.abs(labels_pred - K.variable(y_test)), .5)
print("Accuracy:", K.get_value(K.sum(K.cast(corrects,"int32")))*100/K.int_shape(corrects)[0])


classes = unique(K.variable(y_test))
print("classes:", K.get_value(classes))

# make y_true a boolean vector
y_true = K.equal(y_test, 1.)

# sort scores and corresponding truth values
indices = argsort(labels_pred)
desc_score_indices = K.get_value(K.reverse(indices, 0))

print("desc_score_indices", desc_score_indices)
y_score = K.gather(labels_pred, desc_score_indices)
y_true = K.gather(y_true, desc_score_indices)
y_true = K.cast(y_true, "int64")
print("y_true", K.get_value(y_true))
print("y_score", K.get_value(y_score))

diff = np.diff(K.eval(y_score))
distinct_value_indices = where(K.not_equal(diff, 0))
distinct_value_indices = K.get_value(distinct_value_indices)[0]
print("distinct_value_indices", distinct_value_indices)
threshold_idxs = K.eval(K.concatenate([K.variable(distinct_value_indices),
                                       K.variable(np.array([largest_index]))], 0))


# accumulate the true positives with decreasing threshold
tps = K.get_value(K.gather(K.cumsum(y_true), threshold_idxs))
print("True Positives:", tps)

fps = 1 + threshold_idxs - tps
print("False Positives:", fps)

thresholds = K.get_value(K.gather(y_score, threshold_idxs))
print("Decreasing Threshold:", thresholds)

plot_roc_curve(tps, fps, thresholds)
```
where the `plot_roc_curve` is:
```python
def plot_roc_curve(tps, fps, thresholds):
    optimal_idxs = np.where(np.r_[True,
                                  np.logical_or(np.diff(fps, 2),np.diff(tps, 2)),
                                    True])[0]
    fps = fps[optimal_idxs]
    tps = tps[optimal_idxs]
    thresholds = thresholds[optimal_idxs]

    tps = np.hstack((0, tps))
    fps = np.hstack((0, fps))
    thresholds = np.hstack((thresholds[0] + 1, thresholds))

    fpr = fps / fps[-1]
    tpr = tps / tps[-1]
    plt.figure()
    lw = 2
    plt.plot(fpr, tpr, color='darkorange', lw=lw)
    plt.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver Operating Characteristic')
    plt.show()
    return
```
Calculating ROC is done based on the [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/2a6b3b8f0ab760f370010b5c060e06787d8a2946/sklearn/metrics/_ranking.py) algorithm.

**Comment by byekrang** [comment]
In order to run the same code with TensorFlow we need a few changes:
- getting `unique`, `where` and `argsort` from TensorFlow (they are lazy by default) and calling a session and placeholder
- TensorFlow does not automatically cast types, so a couple of explicit type casting are needed
- TensorFlow's `where` does not result in the dimensionan as NumPy's, which changes the calculation of `distinct_value_indices `
#### `where` in TensorFlow
```python
sess = tf.InteractiveSession()
sess.run(tf.where(tf.not_equal(np.array([1,1,2,3,4,4]), 1)))
```
```pytb
array([[2],
       [3],
       [4],
       [5]])
```
#### `where` in NumPy
```python
np.where(np.array([1,1,2,3,4,4]) != 1)
```
```pytb
(array([2, 3, 4, 5]),)
```
We can write the code as:
```python
from keras import backend as K
from keras.datasets import imdb
import tensorflow as tf
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

unique = tf.unique
argsort = tf.argsort
where = tf.where


max_features = 10000
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
largest_index = y_test.size - 1

df = pd.read_csv('labels_pred.csv')

labels_pred = K.squeeze(df.values, 1)

print("Predicted labels:", K.get_value(labels_pred))
plt.hist(K.get_value(labels_pred))
plt.show()

###
corrects = K.less(K.abs(labels_pred - K.variable(y_test, dtype= "float64")), .5)
print("Accuracy:", K.get_value(K.sum(K.cast(corrects,"int32")))*100/K.int_shape(corrects)[0])

x = K.placeholder(shape=(None))
classes = unique(x)
sess = tf.InteractiveSession()
print("classes:", sess.run(classes[0], {x:y_test}))

# make y_true a boolean vector
y_true = K.equal(K.cast(y_test, "float32"), 1.)

# sort scores and corresponding truth values
indices = argsort(labels_pred)
desc_score_indices = K.get_value(K.reverse(indices, 0))
print("desc_score_indices", desc_score_indices)

y_score = K.gather(labels_pred, desc_score_indices)
y_true = K.gather(y_true, desc_score_indices)
y_true = K.cast(y_true, "int64")
print("y_true", K.get_value(y_true))
print("y_score", K.get_value(y_score))

diff = np.diff(K.eval(y_score))
distinct_value_indices = where(K.not_equal(diff, 0))
distinct_value_indices = K.get_value(K.squeeze(distinct_value_indices, 1))
print("distinct_value_indices", distinct_value_indices)
threshold_idxs = K.eval(K.cast(K.concatenate([K.variable(distinct_value_indices),
                                       K.variable(np.array([largest_index]))], 0),
                              "int64"))


# accumulate the true positives with decreasing threshold
tps = K.get_value(K.gather(K.cumsum(y_true), threshold_idxs))
print("True Positives:", tps)

fps = 1 + threshold_idxs - tps
print("False Positives:", fps)

thresholds = K.get_value(K.gather(y_score, threshold_idxs))
print("Decreasing Threshold:", thresholds)
```
